ocaml/src/proxy_osd.ml

```diff
@@ -21,18 +21,243 @@ open Slice
+        )
+        !endpoints >>= fun resolveds ->
+      Lwt.return (List.flatten_unordered resolveds)
+    in
+    let rec inner () =
+      if !fuse
+      then Lwt.return ()
+      else
+        begin
+          resolve_endpoints () >>= fun resolved_endpoints ->
+
+          Lwt_log.debug_f "proxy_osd: resolved endpoints to: %s"
+                          ([%show : (Net_fd.transport * string * int) list] resolved_endpoints) >>= fun () ->
+
+          let active_endpoints = Hashtbl.keys active_pools in
+          let disqualified_endpoints = Hashtbl.keys disqualified_endpoints_requalify_threads in
+
+          let all_current_endpoints = List.rev_append active_endpoints disqualified_endpoints in
+          let new_endpoints =
+            List.filter
+              (fun endpoint -> not (List.mem endpoint all_current_endpoints))
+              resolved_endpoints
+          in
+
+          List.iter requalify_endpoint new_endpoints;
+
+          let active_to_remove =
+            List.filter
+              (fun endpoint -> not (List.mem endpoint resolved_endpoints))
+              active_endpoints
+          in
+          List.iter
+            (fun endpoint ->
+              let pp = Hashtbl.find active_pools endpoint in
+              pp # finalize |> Lwt.ignore_result;
+              Hashtbl.remove active_pools endpoint)
+            active_to_remove;
+
+          let disqualified_to_remove =
+            List.filter
+              (fun endpoint -> not (List.mem endpoint resolved_endpoints))
+              disqualified_endpoints
+          in
+          List.iter
+            (fun endpoint -> (Hashtbl.find disqualified_endpoints_requalify_threads endpoint |> snd) := true)
+            disqualified_to_remove;
+
+          Lwt_extra2.sleep_approx 60. >>= fun () ->
+          inner ()
+        end
+    in
+    Lwt.ignore_result (inner ())
+  in
+
+  object(self :# proxy_pool)
+    method with_client ~namespace f =
+      if !fuse
+      then Lwt.fail_with "multi_proxy_pool is being finalized"
+      else
+        get_pool ~namespace >>= fun (endpoint, pp) ->
+        Lwt.catch
+          (fun () -> pp # with_client ~namespace f)
+          (fun exn ->
+            requalify_endpoint endpoint;
+            Lwt.fail exn)
+
+    method finalize =
+      Lwt_log.info_f "Finalizing multi_proxy_pool" >>= fun () ->
+      fuse := true;
+      Hashtbl.iter (fun _ p -> p # finalize |> Lwt.ignore_result) active_pools;
+      Hashtbl.clear active_pools;
+      Hashtbl.iter (fun _ (_, fuse) -> fuse := true) disqualified_endpoints_requalify_threads;
+      Lwt.return ()
+  end
+
+let ensure_namespace_exists proxy_client ~namespace ~preset =
+  proxy_client # list_namespaces
+               ~first:namespace ~finc:true
+               ~last:(Some (namespace, true))
+               ~max:1 ~reverse:false
+  >>= fun ((_, x), _) ->
+  match x with
+  | [] -> proxy_client # create_namespace ~namespace ~preset_name:(Some preset)
+  | [ _; ] -> Lwt.return ()
```

Comment:
That would be rather unexpected ...
What would the scenario be where we get into such a situation?
- they deliberately already pre-create this namespace with some other preset
- they call add-osd with preset A, which succeeds in creating the namespace, but fails before adding the osd to the albamgr. Next they call add-osd again with another preset

One is sabotage, the other very unlikely. If you want I can add detection for it and fail with some error, in practice however I don't expect this to be a problem?